DOC Updates for symfony/mailer (#92)
silverstripe/developer-docs: en/02_Developer_Guides/10_Email/index.md

# Email

> Creating and sending email in Silverstripe CMS is done using the [Email](api:SilverStripe\Control\Email\Email). This document covers how to create an `Email` instance and customise it with an HTML template.

Reviewer: ```suggestion
Creating and sending email in Silverstripe CMS is done using the [Email](api:SilverStripe\Control\Email\Email) class. This document covers how to create an `Email` instance and customise it with an HTML template.
```

## Configuration

> Silverstripe CMS provides an API over the top of [symfony/mailer](https://symfony.com/doc/current/mailer.html) which comes with an extensive list of "transports" for sending mail via different services such as SMTP, Gmail and Amazon SES.

Reviewer: ```suggestion
Silverstripe CMS provides an API over the top of [symfony/mailer](https://symfony.com/doc/current/mailer.html) which comes with an extensive list of "transports" for sending mail via different services such as SMTP, Gmail, and Amazon SES.
```

Author: We could argue all day about this https://www.grammarly.com/blog/comma-before-and/

Reviewer: Does that mean you'll make the change to avoid an argument or you won't make the change and I'll just have to live without my oxford comma?

Author: I performed a string search for 
`, and`
in 
`vendor/silverstripe/**/*.md`

There are numerous example of using a comma before `and` in a list, much to my horror :-)

Author: Updated

> Email configuration is done using Symfony's `[DSN](https://symfony.com/doc/current/mailer.html#transport-setup) configuration string which is used to select which transport is used and any required configuration such as username and password. In Silverstripe, this is done with either an environment variable or yml configuration.

Reviewer: ```suggestion
Email configuration is done using Symfony's [DSN](https://symfony.com/doc/current/mailer.html#transport-setup) configuration string which is used to select which transport is used and any required configuration such as username and password. In Silverstripe, this is done with either an environment variable or yml configuration.
```

Author: Updated

> The `Sendmail` transport is the most common one and is used by default in Silverstripe. The `sendmail` binary is widely available across most Linux/Unix servers. By default the sendmail command used is `/usr/sbin/sendmail -bs`.

Reviewer: ```suggestion
The `Sendmail` transport is the most common one and is used by default in Silverstripe. The `sendmail` binary is widely available across most Linux/Unix servers. By default the sendmail command used is `/usr/sbin/sendmail -bs`, but this [can be configured](#dsn-sendmail) as part of the `DSN`.
```

Author: Updated

> Alternatively you can provide a different `DSN` to select any of the Transport classes provided natively by `symfony/mailer` or other compatible third-party transports. For more information and to see what other transports are available see the [symfony/mailer transport types](https://symfony.com/doc/current/mailer.html#using-a-3rd-party-transport).

Reviewer: ```suggestion
Alternatively you can provide a different `DSN` to select any of the Transport classes provided natively by `symfony/mailer` or other compatible third-party transports. For more information and to see what other transports are available see the [symfony/mailer transport types](https://symfony.com/doc/current/mailer.html#using-a-3rd-party-transport).

[hint]
The format for the DSN is exactly as defined in the symfony docs linked above. Some common examples are listed below.
[/hint]
```
Just in case anyone is confused - lets make it super clear that the symfony docs show what the dsn should look like - we're just passing it along.

Author: Updated

To set the DSN string in an environment variable (recommended):

**.env**
```
MAILER_DSN="<my-dsn>"
```

To set the DSN string in project yml:

**app/_config/myemail.yml**
```yml
---
Name: mailer-dsn-config

> After: '*'

Reviewer: Does this need to be after everything?

Author: I think it's safest to. This means it will always have precedence.

Technically it could possibly be `After: 'mailer-dsn-env'` - relates to https://github.com/silverstripe/silverstripe-framework/pull/10494/files

However
- I'm not 100% sure if 'mailer-dsn-env' yml block will always exist, as there's an `only` conditional in there which checks to see the `MAILER_DSN` env variable exists.
- we might need to add in another yml block that goes after 'mailer-dsn-env' at a later date, which would then take precedence over project config

There's a different argument to be made that the env variable should have precedence over any project yml (it's a pretty strong argument, I may have got it the wrong way around when I first though about this).

Probably makes sense to leave that argument for the actual framwork PR though

Should leave this unresolved for now

Reviewer: I agree that this should be changed so that the .env config takes precedence (see https://github.com/silverstripe/silverstripe-framework/pull/10494/files#r985306124). Please update the docs to match.

Author: Have updated docs, note: I've have kept the `MAILER_DSN` config in yml, rather than overriding in php.  This is because the service is created via new injector factory yml config. This also allows more flexibility in the case there's a misconfigured env variable on a prod server that needs to be urgently changed by developers who can't quickly change an env variable (though I haven't mentioned that explicitly in the docs).

I've updated the docs accordingly, specifically mentioning that project config must be exactly naming `mailer-dsn-project-config` to allow the env variable to retain precedence.

---
SilverStripe\Core\Injector\Injector:
  Symfony\Component\Mailer\Transport\TransportInterface:
    constructor:

>       dsn: '<my-dsn>'

Reviewer: The old docs had examples here for authentication - does symfony mailer simply never need authentication (I would find this highly suspect)? If it does, how do we pass that through? Is that all part of the dsn?
If so we should mention that fact above as well.

Author: See below, I provided an example for an SMTP DSN string

Reviewer: In that case:
> we should mention that fact above as well.

Reviewer: Actually nevermind, the example below is close enough to here that it's not needed.

```

> The precedence for configuration, from highest to lowest:

Reviewer: ```suggestion
The configuration priority order is as follows, from highest to lowest:
```

Author: Updated

- Project yml containing `After: '*'`
- The `MAILER_DSN` environment variable
- The default silverstripe DSN of `sendmail://default` which will use `/usr/sbin/sendmail -bs`

### Common DSN strings

> To configure SMTP

Reviewer: ```suggestion
#### To configure SMTP {#dsn-smtp}
```

Author: Updated

**.env**
```
MAILER_DSN="smtp://user:pass@smtp.example.com:1234"
```

> To configure a different sendmail binary and command

Reviewer: ```suggestion
#### To configure a different sendmail binary and command {#dsn-sendmail}
```

Author: Updated

**.env**
```
MAILER_DSN="sendmail://default?command=/path/to/mysendmailbinary%20-t"
```

> To suppress all emails

Reviewer: ```suggestion
#### To suppress all emails {#dsn-null}
```

Author: Updated

**.env**
```
MAILER_DSN="null://default"
```

> Read more about other [DSN strings available](https://symfony.com/doc/current/mailer.html#using-a-3rd-party-transport)

Reviewer: ```suggestion
Read more about other available DSN strings in [the symfony documentation](https://symfony.com/doc/current/mailer.html#using-a-3rd-party-transport)
```

Author: Updated

# Email

> Creating and sending email in Silverstripe CMS is done using the [Email](api:SilverStripe\Control\Email\Email). This document covers how to create an `Email` instance and customise it with an HTML template.

Reviewer: ```suggestion
Creating and sending email in Silverstripe CMS is done using the [Email](api:SilverStripe\Control\Email\Email) class. This document covers how to create an `Email` instance and customise it with an HTML template.
```
as per https://github.com/silverstripe/developer-docs/pull/92#discussion_r983053035

Author: Updated

### Sending plain text only

```php
use SilverStripe\Control\Email\Email;

> $email = new Email('from@mydomain.com', 'to@example.com', 'My subject', 'My email body text');

Reviewer: I prefer the old example - most of the time people _will_ already have these in variables, so it's easy to just copy paste that into their projects, and the variables are self explanatory anyway.

Author: Updated

from the system by stripping HTML markup, or transforming it where possible (e.g. `<strong>text</strong>` is converted 
to `*text*`).

```php
use SilverStripe\Control\Email\Email;

$email = new Email('from@mydomain.com', 'to@example.com', 'My subject', 'My email body text');
$email->send();

> ```

Reviewer: ```suggestion
```
This example is superfluous - it's already shown in the "Testing how email works" section which isn't far above here. At most the paragraph above could say "By default, when using the `send()` method" at the start.

Author: Removed

> You can also specify separate plain text content so that it's not automatically generated from HTML

Reviewer: ```suggestion
You can also specify plain text and HTML content separately if you don't want the plain text to be automatically generated from HTML
```

Author: Updated

```php
use SilverStripe\Control\Email\Email;

> $email = new Email('from@mydomain.com', 'to@example.com', 'My subject');

Reviewer: ```suggestion
$email = new Email($from, $to, $subject);
```

Author: Updated

$email->html('<p>My HTML email content</p>');
$email->text('My plain text email content');
$email->send();
```

[info]
The default HTML template for emails is `vendor/silverstripe/framework/templates/SilverStripe/Control/Email/Email.ss`.

> To customise this template, copy it to the `app/templates/Email/` folder or use `setHTMLTemplate` when you create the 

Reviewer: Have you tested this? I think it would have to be in the `app/templates/SilverStripe/Control/Email/` folder.
Note: I haven't tested it yet - this first review is just a first pass finding what I can visually first.

Author: Have updated instructions to say copy it to `<project-root>/themes/<my-theme>/SilverStripe/Control/Email/Email.ss` - I just tested this works out of the box without needing to specify the location with `setHTMLTemplate()`

Reviewer: Yeah `setHTMLTemplate()` is for setting a specific template (e.g. if you have different templates for different emails)

> To customise this template, copy it to the `app/templates/Email/` folder or use `setHTMLTemplate` when you create the 

Reviewer: ```suggestion
To customise this template, copy it to the `app/templates/Email/` folder, or use `setHTMLTemplate()` to use a template in a different location when you create the 
```

Author: Updated slightly differently based on change above

`Email` instance. Note - by default the `$EmailContent` variable will escape HTML tags for security reasons. If you feel confident allowing this variable to be rendered as HTML, then update your custom email template to `$EmailContent.RAW`
[/info]

### Templates

HTML emails can use custom templates using the same template language as your website template. You can also pass the
email object additional information using the `setData` and `addData` methods. 

**app/templates/Email/MyCustomEmail.ss**

> ```html

Reviewer: Revert this change. `ss` is correct - docs.silverstripe.org handles it correctly.
https://docs.silverstripe.org/en/4/developer_guides/email/#templates

Author: Reverted

<h1>Hi $Member.FirstName</h1>
<p>You can go to $Link.</p>
```

The PHP Logic..

```php

use SilverStripe\Control\Email\Email;

$email = Email::create()
    ->setHTMLTemplate('Email\\MyCustomEmail') 
    ->setData([
        'Member' => Security::getCurrentUser(),
        'Link'=> $link,
    ])

>     ->from('from@mydomain.com')
>     ->to('to@example.com')
>     ->subject('My subject');

Reviewer: ```suggestion
    ->from($from)
    ->to($to)
    ->subject($subject);
```

Author: Updated

the plain email:

```php
use SilverStripe\Control\Email\Email;

> $email = new Email('from@mydomain.com', 'to@example.com', 'My subject', 'My email body text');

Reviewer: ```suggestion
$email = new Email($from, $to, $subject, $body);
```

Author: Updated

You can set the default sender address of emails through the `Email.admin_email` [configuration setting](/developer_guides/configuration).

**app/_config/app.yml**

> ```yml

Reviewer: Do you know for a fact this will format correctly? The existing `yaml` is definitely formatting correctly so I'm hesitant to change this. Probably best to leave these alone for now - there's [an existing card](https://github.com/silverstripe/developer-docs/issues/11) to standardise this sort of thing so if we want to standardise `yaml` to `yml` we can do it as part of that card.
I've added standardisation of `yaml` or `yml` to the card's proposed changes.

Author: Reverted

marked as spam.

```php
use SilverStripe\Control\Email\Email;

> $email = new Email('from@mydomain.com', 'to@example.com', 'My subject', 'My email body text');

Reviewer: ```suggestion
$email = new Email($from, $to, $subject, $body);
```

Author: Updated

$email->replyTo('reply@example.com');
```

## Advanced customisation

> Silverstripe Email is built on top of symfony/mailer for for advanced customisation information refer to the [symfony/mailer docs](https://symfony.com/doc/current/mailer.html)

Reviewer: ```suggestion
Silverstripe Email is built on top of [symfony/mailer](https://github.com/symfony/mailer). For advanced customisation information refer to the [symfony/mailer docs](https://symfony.com/doc/current/mailer.html)
```

Author: Updated
